### PR TITLE
Feat/dpendency validation

### LINKED ISSUE #855: 'validate' to identify incorrect transitive dependencies
sfdx-project.json is the manifest which determines the packages in a particular project, dependencies of each packages and deployment order. As projects grow, the complexity of this manifest grows (3K+ lines in some projects) making it incomprehensible. 

To simplify the management of above and prevent errors be detected late in the cycle , this work item aims to do the following
**Using a linter like approach, break validate and provide warning to the user on unlocked packages missing transitive dependencies, error such as this are currently only detected during build stage**

This can be based on the work done by @Rocko1204 in https://github.com/Rocko1204/packaging-helper-cli

Refer to https://github.com/dxatscale/dxatscale-issue-registry/issues/18 for motivation

### PR DESCRIPTION
**Pull request by zhebinliu**
#### Checklist
All items have to be completed before a PR is merged

- [x] Adhere to [Contribution Guidelines](https://docs.dxatscale.io/about-us/contributing-to-dx-scale)
- [ ] Updates to Decision Records considered?
- [ ] Updates to documentation at [DX@Scale Guide](https://github.com/dxatscale/dxatscale-guide) considered?
- [x] Tested changes?
- [x] Unit Tests new and existing passing locally?

### CODE REVIEW

File: packages/core/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@dxatscale/sfpowerscripts.core",
-    "version": "32.0.9",
+    "version": "32.0.10",
```
Comment (azlam-abdulsalam) [inline packages/core/package.json]: lets remove versioning using PRs. and this should be a minor as well

File: packages/core/src/dependency/TransitiveDependencyResolver.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    this.dependencyMap = await this.getAllPackageDependencyMap(this.projectConfig);
+
+    await this.validateDependency(this.dependencyMap, this.projectConfig)
+
+    return this.updatedprojectConfig
+     
+  }
+
+  private getAllPackageDependencyMap(projectConfig: any): any {
```
Comment (azlam-abdulsalam) [inline packages/core/src/dependency/TransitiveDependencyResolver.ts]: define type

File: packages/core/src/dependency/TransitiveDependencyResolver.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        pkgWithDependencies[pkg.package] = pkg.dependencies
+      }
+    }
+    return pkgWithDependencies
+  }
+
+  private async validateDependency(dependencyMap: any, projectConfig: any){
+    let pkgs = Object.keys(dependencyMap)
+    for ( let pkg of pkgs){
+      console.log(
```
Comment (azlam-abdulsalam) [inline packages/core/src/dependency/TransitiveDependencyResolver.ts:47]: Replace console logs with SFPLogger and adequate loglevels  also look into the targer such as file or console

File: packages/core/src/dependency/TransitiveDependencyResolver.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      ];
+
+      table.push(item);
+    }
+
+    return table;
+
+  }
+
+  private async ExternalPackageDependencyFetcher(packageId: string){
```
Comment (azlam-abdulsalam) [inline packages/core/src/dependency/TransitiveDependencyResolver.ts:101]: function name typo

File: packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    conn = await setupFakeConnection();
+
+  })
+
+  it("should resolve missing package dependencies with transitive dependency", async () => {
+    const transitiveDependencyResolver = new TransitiveDependencyResolver(projectConfig, conn);
+    const resolvedProjectConfig: any = await transitiveDependencyResolver.exec();
+
+    let packageDescriptor = resolvedProjectConfig.packageDirectories.find((dir) => dir.package === "candidate-management");
+    let coreDependency = packageDescriptor.dependencies.find(dependency => dependency.package === "temp");
```
Comment (azlam-abdulsalam) [inline packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts]: I didn't understand what the test does here,

File: packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    expect(coreDependency.versionNumber).toBe("1.0.0.LATEST");
```
Comment (azlam-abdulsalam) [inline packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts]: Is it wrong expect, shouldn't be the test whether temp got added

File: packages/sfpowerscripts-cli/src/impl/parallelBuilder/BuildImpl.ts [hunk 7]
```diff
@@ -86,16 +88,18 @@ export default class BuildImpl {
     }> {
         if (this.props.devhubAlias) this.sfpOrg = await SFPOrg.create({ aliasOrUsername: this.props.devhubAlias });
 
+        //Validate dependencies in sfdx-project.json // Read Manifest
+        this.projectConfig = ProjectConfig.getSFDXProjectConfig(this.props.projectDirectory);
+        const transitiveDependencyResolver = new TransitiveDependencyResolver(this.projectConfig, this.sfpOrg.getConnection());
```
Comment (azlam-abdulsalam) [inline packages/sfpowerscripts-cli/src/impl/parallelBuilder/BuildImpl.ts]: Group it as a function and move it after doing git operations

File: packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  return GitTags;
+});
+
+let conn: Connection;
+let gitTags;
+let response;
+
+describe("Given a TransitiveDependencyResolver", () => {
+
+  beforeEach(async () => {
```
Comment (azlam-abdulsalam) [inline packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts:52]: Being a very core feature, can we add.more tests, such as ordering?

**Review by azlam-abdulsalam — CHANGES_REQUESTED** [review]
Please see my comments

File: packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts (hunk 5 above)
Comment (zhebinliu) [reply packages/core/tests/package/dependencies/TransitiveDependencyResolver.test.ts]: This is to check whether the missing dependencies are added, and the version is same as its dependent pacakge

**Pull request closed without merging**